src/compiler/checker.ts

```diff
@@ -4723,21 +4726,32 @@ module ts {
 
         function getInferredType(context: InferenceContext, index: number): Type {
             let inferredType = context.inferredTypes[index];
+            let inferenceSucceeded: boolean;
             if (!inferredType) {
                 let inferences = getInferenceCandidates(context, index);
                 if (inferences.length) {
                     // Infer widened union or supertype, or the undefined type for no common supertype
                     let unionOrSuperType = context.inferUnionTypes ? getUnionType(inferences) : getCommonSupertype(inferences);
-                    inferredType = unionOrSuperType ? getWidenedType(unionOrSuperType) : inferenceFailureType;
+                    inferredType = unionOrSuperType ? getWidenedType(unionOrSuperType) : unknownType;
+                    inferenceSucceeded = !!unionOrSuperType;
                 }
                 else {
                     // Infer the empty object type when no inferences were made
                     inferredType = emptyObjectType;
+                    inferenceSucceeded = true;
                 }
-                if (inferredType !== inferenceFailureType) {
+
+                // Only do the constraint check if inference succeeded (to prevent cascading errors)
+                if (inferenceSucceeded) {
                     let constraint = getConstraintOfTypeParameter(context.typeParameters[index]);
                     inferredType = constraint && !isTypeAssignableTo(inferredType, constraint) ? constraint : inferredType;
                 }
+                // If inference failed, it is necessary to record the index of the failed type parameter (the one we are on).
+                // It might be that inference has already failed on a later type parameter on a previous call to inferTypeArguments.
+                // So if this failure is on preceding type parameter, this type parameter is the new failure index.

```

Comment:
I'd prefer if we don't put comments before 'else'.  it makes the 'if' portion appear 'done', and makes it easy to miss htat this side is an 'else' vs just hte code that runs once the 'if' is done.  Can we put the comments 'inside' the 'else' block instead?